File: src/modules/algorithm/seed/algorithms/queue/languages/typescript/array-solution.ts
export class Queue {
  private items: any[];
  private capacity: number;
  private front: number;
  private rear: number;
  private count: number;

  constructor(capacity: number) {
    this.items = new Array(capacity);
    this.capacity = capacity;
    this.front = 0;
    this.rear = -1;
    this.count = 0;
  }

  enqueue(item: any): void {
    if (this.isFull()) {
      throw new Error('Queue Overflow');
    }
    this.rear = (this.rear + 1) % this.capacity;
    this.items[this.rear] = item;
    this.count++;
  }

  dequeue(): any {
    if (this.isEmpty()) {
      return undefined;
    }
    const item = this.items[this.front];
    this.front = (this.front + 1) % this.capacity;
    this.count--;
    return item;
  }

  peek(): any {
    if (this.isEmpty()) {
      return undefined;
    }
    return this.items[this.front];
  }

  isEmpty(): boolean {
    return this.count === 0;
  }

  isFull(): boolean {
    return this.count === this.capacity;
  }

  size(): number {
    return this.count;
  }
}

File: src/modules/algorithm/seed/algorithms/ring-buffer/languages/go/array-solution.go
package ringbuffer

// RingBuffer represents a circular buffer with a fixed capacity.
type RingBuffer struct {
	buffer   []interface{}
	head     int
	tail     int
	count    int
	capacity int
}

// NewRingBuffer creates a new RingBuffer with the given capacity.
func NewRingBuffer(capacity int) *RingBuffer {
	return &RingBuffer{
		buffer:   make([]interface{}, capacity),
		head:     0,
		tail:     0,
		count:    0,
		capacity: capacity,
	}
}

// Enqueue adds an element to the ring buffer.
func (rb *RingBuffer) Enqueue(item interface{}) {
	rb.buffer[rb.head] = item
	rb.head = (rb.head + 1) % rb.capacity
	if rb.count == rb.capacity {
		rb.tail = (rb.tail + 1) % rb.capacity
	} else {
		rb.count++
	}
}

// Dequeue removes and returns the oldest element from the ring buffer.
// Returns nil if the buffer is empty.
func (rb *RingBuffer) Dequeue() interface{} {
	if rb.IsEmpty() {
		return nil
	}
	item := rb.buffer[rb.tail]
	rb.buffer[rb.tail] = nil // Avoid memory leaks
	rb.tail = (rb.tail + 1) % rb.capacity
	rb.count--
	return item
}

// IsFull returns true if the ring buffer is full.
func (rb *RingBuffer) IsFull() bool {
	return rb.count == rb.capacity
}

// IsEmpty returns true if the ring buffer is empty.
func (rb *RingBuffer) IsEmpty() bool {
	return rb.count == 0
}

// GetCount returns the number of elements currently in the ring buffer.
func (rb *RingBuffer) GetCount() int {
	return rb.count
} 